Improved groups search for LDAP login

LeonDaniel opened the pull request:
- added another method to search LDAP groups by querying the userDN for 'memberOf' attribute

jefferai commented on builtin/credential/ldap/backend.go:
```diff
@@ -139,13 +143,27 @@ func (b *backend) Login(req *logical.Request, username string, password string)
+						if err != nil || len(memberOfDN.RDNs) == 0 || len(memberOfDN.RDNs[0].Attributes) == 0 {
+							continue
+						}
+						// I assume the standard states that CN is the first RDN attribute
```
Do you have a reference for this?

jefferai commented:
I think it looks pretty good, but it'd be nice to see a formal reference for that assumption you made, otherwise it should be checked to be sure.

LeonDaniel commented:
I searched in RFC's for this, but it seems that this is an operational attribute and not mandatory. The 'memberOf' attribute is implementation specific in Microsoft Active Directory (https://msdn.microsoft.com/en-us/library/windows/desktop/ms677099%28v=vs.85%29.aspx)
Using my company AD, this was the only way I could retrieve the LDAP groups, as member attribute (mentioned in RFC 4519 section 2.17) was not set.
I kept the standard way of retrieving the groups through 'member', 'uniqueMember' and 'memberUid' and I appended any different groups found within 'memberOf' attribute, so it should be backward compatible if someone is using a RFC compliant LDAP server.

jefferai commented:
Hi @LeonDaniel ,

AD not following the spec is old news...heh. What I was actually referring to, though, was whether it's a valid assumption that `cn` will always be the first attribute returned, so whether your assumption that you can always `[0]` is correct, or whether they should be iterated through. If there's no documentation supporting that assumption, I'd prefer to see iteration and checking the parameters to be on the safe side.

LeonDaniel commented:
Your concern makes perfect sense. I used it this way (along with that suspicious comment) because the way of retrieving the groups from groupDN searching for 'member' & co makes the same assumption that CN is the first RDN (lines 176 to 183); lines which where commited before my changes.
I can iterate and search directly for CN and I think I should make the same for the search within groupDN. Should this suffice ?

jefferai commented:
I didn't notice that -- that's sadmaking. (And it's actually possible that that's related to an issue someone else is having.) Fixing both would be fantastic!

LeonDaniel commented:
OK. I'll commit the changes is a few minutes.

LeonDaniel commented:
Sorry for the delay. I had to test the changes. Please review when you have a chance. Thanks!

LeonDaniel commented:
Is there any chance that this change will end up in a release version any time soon ?
I would like to use a official version, instead of my custom made executable.

jefferai commented:
Hi @LeonDaniel ,

We just released 0.5.2; I don't yet know if the next version is 0.5.3 or 0.6 (probably 0.6) but it's not planned to be in the very near term as we have a lot to do beforehand. Sorry...

jefferai commented on builtin/credential/ldap/backend.go:
```diff
@@ -155,23 +180,33 @@ func (b *backend) Login(req *logical.Request, username string, password string)
+			return nil, nil, err
+		}
+		for _, rdn := range dn.RDNs {
+			for _, rdnTypeAndValue := range rdn.Attributes {
```
I get what you're doing here...my question is, are you sure that "CN" is the type for group values across all LDAP servers?

jefferai commented:
Hi Leon,

I made one comment on the code, and there is a merge conflict. Once you address those I'll merge in!

LeonDaniel commented:
The CN attribute from the RDN is the one that defines the standard name of a LDAP object.(https://tools.ietf.org/html/rfc4519#section-2.3). 
And is still a better option than relying on the first value of the RDNs(method that was still used when I merged). 
If you want a reassurance for each LDAP server there is, I don't thing that I can give you because I don't know if this attribute is mandatory or not in their implementation.
As a example openLDAP suggests to use UID instead of CN in a object DN in their documentation (but the default is CN) and I can't find anything on the memberOf if it could ever have uid instead of CN.
As of Active Directory, well...basically you can do whatever you want.
I found a lot of reference that suggests that CN is pretty standard when in comes to defining a LDAP object name in many forums / stackoverflow posts. I can search for more references if you want.

jefferai commented:
Nope, thanks for the clarification. Any chance you and Steve from #1335 can coordinate on a set of changes that works for both of you?

LeonDaniel commented:
I made some changes in the latest commit that I hope can work for both of us. When you have some time, please review them.

jefferai commented on builtin/credential/ldap/backend.go:
```diff
@@ -101,106 +102,191 @@ func (b *backend) Login(req *logical.Request, username string, password string)
 	}
-	binddn := ""
+
+	bindDN, response := getBindDN(cfg, c, username)
```
I really like that you split this logic out into different functions, but please follow the convention of returning an `error` rather than returning a `*logical.Response`.

jefferai commented on builtin/credential/ldap/backend.go:
```diff
@@ -101,106 +102,191 @@ func (b *backend) Login(req *logical.Request, username string, password string)
+	if err != nil {
+		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP fetch of distinguishedName=%s failed: %v", userdn, err))
+	}
+	if len(result.Entries) != 1 {
```
What happens if there is no `memberOf` attribute? I would assume that in that case either `err` is not `nil` or `result.Entries` is empty, but it looks like if either of those is true, the function will exit before running the other searches later on.

jefferai commented on builtin/credential/ldap/backend.go:
```diff
@@ -101,106 +102,191 @@ func (b *backend) Login(req *logical.Request, username string, password string)
+		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP fetch of distinguishedName=%s failed: %v", userdn, err))
+	}
+	if len(result.Entries) != 1 {
+		return nil, logical.ErrorResponse("LDAP search for binddn 0 or not unique")
```
This isn't searching for `binddn` :-)

jefferai commented:
Made a few comments, looking really good to me, otherwise!

LeonDaniel commented:
Fixed. I wanted to return an error, but the go tutorials suggested creating a new type and attaching the Error method to implement the interface, so I took a shortcut :)) . Sorry if the code is not the best, but I learned Go in about 2 weeks just for fun and started using it here.
I removed the check for the length of the results from the userDN query, somehow I missed that. Thanks !

jefferai commented:
There are definitely reasons why and when you might want a new type with an error method, but I think in that case it was mostly a Go tutorial just trying to make you aware that you *can* do that. In this case I'm definitely valuing convention and consistency  :-)

Hopefully @steve-jansen can take this for a test-drive soon.

jefferai commented on builtin/credential/ldap/backend.go:
```diff
@@ -193,61 +193,61 @@ func getBindDN(cfg *ConfigEntry, c *ldap.Conn, username string) (string, *logica
-		return nil, logical.ErrorResponse("LDAP search for binddn 0 or not unique")
+		return nil, fmt.Errorf("LDAP fetch of distinguishedName=%s failed: %v", userDN, err)
 	}
+	// if there are more than one entry, we consider the results irrelevant and ignore them
```
Would the more than one entry mean we found more than one user with the same `cn` at the given `userDN`? If so, probably we should error in this case, right? I worry about silently discarding bad results.

LeonDaniel replied:
No, the search is directly on the userDN object, not in the child tree like the others. It's like searching for the primary key and theoretically there shouldn't be more than one ldap object with our detected userDN. @steve-jansen made this check just to be absolutely sure in case there could be objects in ldap server that have the same DN(but improbable).

LeonDaniel replied:
But your reasoning is right. If I look into the getUserDN function when upnDomain is not empty I set as the found userDN the DN of the last entry from the search results. And there should be only one entry here. I think I should add a check for uniqueness here also like in the bindDN discovery.
That's what is happening when you rely on the old code as a guideline.

LeonDaniel replied:
I also have to check if the results from a ldap query contains a error different than nil if the result list is empty. If that's the case, I should remove the checks for nil in the getLdapGroups function as well.
I really could use a ldap server that doesn't have memberOf in the userDN objects, so I can test the theory. That should take a while as I will have to install an OpenLDAP server in a docker instance or vm.

jefferai replied:
> That's what is happening when you rely on the old code as a guideline.

Yeah, the LDAP backend code isn't the greatest. We had it slated for a rewrite but then people kept sending over PRs until the really bad things were basically solved, so a full rewrite has fallen off the radar. Efforts like yours are *fabulous* and very much appreciated.

LeonDaniel commented:
I will make the final commit next week, so I can have more time to investigate properly. Do you think  this PR will be merged into the next version ? Should I hurry up ?

jefferai commented:
I think it will be merged in if it's ready before we release. In a perfect world that probably/hopefully means that you should finish this by early next week.

jefferai commented:
Hi @LeonDaniel ,

Just wondering how things are going!

LeonDaniel commented:
Hello,

I will commit the changes after finishing the tests later today.

steve-jansen commented:
@LeonDaniel did you find a publicly accessible LDAP server that uses `memberOf`?  Doesn't look like the ldap.forumsys.com test server will help here.

jefferai commented:
My rather naive understanding is that basically any AD server should return memberOf...no?

LeonDaniel commented:
@steve-jansen I am using a openldap docker image and adding the entries the way I want.
@jefferai I don't know for sure if every AD has this member. I found that is implemented in my company AD and is the only way to retrieve the member groups. I suppose that is the default way as the guys from the system didn't tell me that they configured it explicitly over default.

LeonDaniel commented:
I made the proper tests and everything seems to be working. In the meantime, I found more information about user groups retrieval.
First : The userPrincipalName exists in AD only( https://ldapwiki.willeke.com/wiki/UserPrincipalName). Of course you could update the OpenLDAP schema to include this member but this is not something that comes out of the box, so we can assume that userPrincipalName is AD specific.
If we look into the method that retrieves the userDN from the configuration userDN, then we see there if you would use a OpenLDAP server, the userDN would not be found, so I would add this into the documentation : If you use AD, then you can add into ldap configuration the upndomain because it will search for the userDN after userPrincipalName; if this attribute does not exist in your user entry do NOT add upndomain to the configuration or it will fail.(I left it blank and updated userattr in my tests with openldap)
As a precaution, I would make some improvements to the getUserDN function(buty I don't thing we still have time) to care into consideration if userPrincipalName is not present to search for the userDN with another filter.
Second: MemberOf is an operational attribute that is found mostly in AD implementation and is maintained by the AD server when the DN of the values changes : https://msdn.microsoft.com/en-us/library/ms676913(v=vs.85).aspx
Third: Member is a more standard attribute defined in https://tools.ietf.org/html/rfc4519 ,
but querying it will return only the first level of groups. If those groups are members of other groups, only one query is not enough. So an improvement is required in the getLdapGroups function also.
Oracle documentation regarding LDAP is pretty interesting:
https://docs.oracle.com/cd/E26217_01/E35769/html/ldap-filters-attrs-users.html
 https://docs.oracle.com/cd/E23824_01/html/821-1455/gladg.html

LeonDaniel commented:
I made my tests with this simple openldap docker image :
https://github.com/osixia/docker-openldap

Don't forget to start the container with port forwarding so you can update the ldap entries:
docker run --name my-openldap-container -p 389:389 --detach osixia/openldap:1.1.2

LeonDaniel commented:
If the release is soon, I would add these improvements as you have a better chance of using ldap if you make the proper configurations. Before this, if someone was using AD, it had no chance whatsoever to use this backend.
I would make 2 more improvements to be completely reliable :
1. make the userDN retrieval independent of userPrincipalName attribute search
2. in the groups search from the configured groupDN, make a recursive call to retrieve nested member groups.
Honestly, I would rewrite the ldap backend as some of the variables from the configuration have a very... heuristic way of being used to login. For example in the documentation userattr is encouraged to be completed with sAMAccountName. If I were setting that value when testing with openLDAP, then when detecting binddn I had no way of getting it right. Maybe there are cases where this should be used that way, but I usually disapprove of multiple meaning variables.

steve-jansen commented:
I'm a big +1 for using Docker images for acceptance tests. Nice idea @LeonDaniel 
@jefferai are you ok with that dependency?

steve-jansen commented:
> Don't forget to start the container with port forwarding so you can update the ldap entries:
docker run --name my-openldap-container -p 389:389 --detach osixia/openldap:1.1.2

@LeonDaniel have you thought about setting up the test dependency in the `PreCheck` block using the [Docker engine SDK for golang](https://github.com/docker/engine-api), along with a `Teardown` block to clean up?

LeonDaniel commented:
Honestly, I'm quite a newbie regarding go programming, and after taking a look into the vault code I saw the PreCheck and Teardown are used in a go test just like startup and teardown methods in junit tests. I should look for examples in the code with this functionality implemented when I have time.
It seems a good idea, but the test environment must have the docker deamon up and running and the user that runs the tests should be part of the docker group to avoid running under root.
Otherwise, using a container as a mocking environment seems a very good way of testing a real scenario. Also custom containers over that base image should be created for adding some cases for users with memberOf attribute, a image for member attribute in group DN, one for both and one for none.

jefferai commented:
Hi all,

We have a little time, we're not going to release today and probably not next so we can merge more in. There's no real time pressure other than my own opinion :-) . So feel free to keep enhancing. If you would like to rewrite the backend, sure... but maybe not for this next release :-)

As for Docker it's been a ToDo for a long while to make the acceptance tests better and Docker was to be a major part of that but we haven't managed to get to it yet. However, Terraform recently switched up their acceptance tests in this way so before putting in a system throughout Vault I'd want to see how much could be reused.

shomron commented on builtin/credential/ldap/backend.go:
```diff
@@ -101,106 +102,191 @@ func (b *backend) Login(req *logical.Request, username string, password string)
 			BaseDN: cfg.GroupDN,
 			Scope:  2, // subtree
-			Filter: fmt.Sprintf("(|(memberUid=%s)(member=%s)(uniqueMember=%s))", ldap.EscapeFilter(username), ldap.EscapeFilter(userdn), ldap.EscapeFilter(userdn)),
+			Filter: fmt.Sprintf("(|(memberUid=%s)(member=%s)(uniqueMember=%s))", ldap.EscapeFilter(username), ldap.EscapeFilter(userDN), ldap.EscapeFilter(userDN)),
```
This line is problematic for Active Directory. By default, the `member` attribute only returns direct members of the group. In our case, we are trying to retrieve all the transitive groups a user is a member of. This can be accomplished with the following query:

`fmt.Sprintf( "(&(objectClass=group)(member:1.2.840.113556.1.4.1941:=%s))", ldap.EscapeFilter(userDN) ) `

I do not know what other servers would do with this. As a result, does it make sense to expose the templated query as configuration, rather than trying to be fancy here? We could have sensible defaults, but I'm not sure that we'll solve for every possible LDAP server.

What are your thoughts? @LeonDaniel @jefferai

jefferai replied:
I think that is a fine idea for when someone has time. In the interim adding your filter to the OR'd list might serve.

steve-jansen replied:
I believe `memberOf` solves this problem.

shomron replied:
@steve-jansen negative. According to [Microsoft's documentation](https://msdn.microsoft.com/en-us/library/ms677943.aspx):
>Be aware that this attribute lists the groups that contain the user in their member attribute—it does not contain the recursive list of nested predecessors. For example, if user O is a member of group C and group B and group B were nested in group A, the memberOf attribute of user O would list group C and group B, but not group A.

I can also confirm that querying the `memberOf` attribute for my user returns _20_ results - my direct group memberships, while using the `LDAP_MATCHING_RULE_IN_CHAIN` I referenced earlier returns **120** groups - direct + transitive membership.

shomron replied:
@jefferai I have time / need to fix this to be able to adopt Vault, so I can take a stab at it - would exposing the configuration as a Go template be agreeable to you? I would merge the user-configured template with a context struct containing escaped `username` and `userdn` to build the query. If so, I'll get cracking on that. Also assuming I should base off of this PR?

steve-jansen commented:
Confirmed this PR correctly implements `memberOf` support for my use case 👍   

I've closed #1294 in favor of this one.

I vote to add tests to this PR to help make the cut-off for the upcoming 0.6.0 release.

Templated ldap queries can be a follow-on release.

jefferai commented:
I think it's certainly a possibility to merge this for 0.6 and look at #1388 for the release after. @shomron would that work for you? That would also make it easy to split the PRs up logically.

shomron commented:
@jefferai that works for me - they are logically separate so it's better to decouple. I'll rebase https://github.com/hashicorp/vault/pull/1388 once you've merged this one and continue with some refinements.

LeonDaniel commented:
If there is something else I can help you with, please let me know.

steve-jansen commented:
@LeonDaniel we need to figure out how to test this in a repeatable way, either unit or acceptance tests.  I'd be very uncomfortable with uncovered blocks that assign authorization groups/roles to authenticated users.

steve-jansen commented:
> However, Terraform recently switched up their acceptance tests in this way so before putting in a system throughout Vault I'd want to see how much could be reused.

I can't find this change in the Terraform repo.  Would you please point me in the right direction?

jefferai commented:
Hi @steve-jansen ,

I'd heard about it but hadn't gotten the details. Basically it's Docker but really Docker via Travis and unfortunately the repo it's in is private, I think because of the need to have keys for various resources stored persistently in order to run the tests.

That would be nice for Vault too, but for local testing it obviously isn't available.

In the past when I was thinking about this it seemed like there were three options:

1. Have a framework/functions to toggle Docker directly within the Vault testing framework
2. Find a library or framework that can do much of the heavy lifting for us
3. See what the Terraform team landed on

So we know the answer to number 3; I haven't been able to do a deep dive into number 2; number 1 is obviously possible but is way outside of the scope of this PR. At the moment we might have to rely on manual testing from trusted community members...but if anyone wants to step up and help figure out the best way forward I'm more than happy to work through it with them.

jefferai commented:
So, given ^, and the choice to defer #1388 until the next release, are we all agreed that this seems ready for a final review and merge?

steve-jansen commented:
@jefferai.  I'm good for merge, with a small nag of discomfort about an authentication backend with untested blocks.  Is there an easy way to mark an automated acceptance test as optional/ignored/disabled by default?  Thinking we could have a Makefile target that does a `docker-compose up` and then runs a few optional tests for the changes in this PR against the running container.

jefferai commented:
We don't have customized Makefile targets for any tests...

jefferai commented:
(Sorry, in case that wasn't clear, it was meant as "we'd want to fit this into the standard Go testing framework".)

shomron commented:
Has anyone looked at [dockertest](https://github.com/ory-am/dockertest) or [go-compose](https://github.com/ibrt/go-compose) for this?

jefferai commented:
Yes.

jefferai commented:
I'm going to merge this - we'll plan the enhancements proposed by @shomron for 0.6.1.
